File: Semester 4/Medii de proiectare si programare/Laborator/Laborator 5/FestivalMuzicaCSharpWFA/FestivalMuzicaCSharpWFA/Service/AuthService.cs
using System.Configuration;
using System.Security.Cryptography;
using System.Text;
using FestivalMuzicaCSharp.Domain;
using FestivalMuzicaCSharp.Repository;

namespace FestivalMuzicaCSharpWFA.Service;

public class AuthService
{
    private static readonly AuthService _instance = new AuthService();
    public static AuthService Instance => _instance;
    private readonly IUsersRepository _repository;
    private readonly string _encryptionAESKey;

    private AuthService()
    {
        var props = new Dictionary<string, string>();
        var connStr = ConfigurationManager.ConnectionStrings["musicFestivalDB"]?.ConnectionString;
        if (connStr == null)
            throw new Exception("Cannot find connection string 'musicFestivalDB' in app.config");
        props["ConnectionString"] = connStr;

        // Load encryption key from config (add this to app.config if not present)
        _encryptionAESKey = ConfigurationManager.AppSettings["encryptionAESKey"];
        if (string.IsNullOrEmpty(_encryptionAESKey))
            throw new Exception("Missing encryptionAESKey in app.config <appSettings>");

        _repository = new UsersDbRepository(props);
    }

    public User Login(string email, string password)
    {
        string encryptedPassword = Encrypt(password, _encryptionAESKey);
        
        var user = _repository.FindByEmailAndPassword(email, encryptedPassword);
        if (user != null)
            return user;
        
        throw new Exception("Invalid credentials");
    }

    public void Register(string email, string password)
    {
        string encryptedPassword = Encrypt(password, _encryptionAESKey);
        
        var existingUser = _repository.FindByEmail(email);
        if (existingUser != null)
            throw new Exception("User already exists with this email");
        
        _repository.Save(new User(0, email, encryptedPassword));
    }

    public string Encrypt(string data, string key)
    {
        try
        {
            using (Aes aesAlg = Aes.Create())
            {
                aesAlg.Key = Encoding.UTF8.GetBytes(key.PadRight(32).Substring(0, 32));
                aesAlg.Mode = CipherMode.ECB;
                aesAlg.Padding = PaddingMode.PKCS7;
                using (ICryptoTransform encryptor = aesAlg.CreateEncryptor())
                {
                    byte[] inputBytes = Encoding.UTF8.GetBytes(data);
                    byte[] encrypted = encryptor.TransformFinalBlock(inputBytes, 0, inputBytes.Length);
                    return Convert.ToBase64String(encrypted);
                }
            }
        }
        catch
        {
            throw new Exception("Password encryption failed!");
        }
    }
}

File: Semester 4/Medii de proiectare si programare/Laborator/Laborator 4/FestivalMuzicaJava/src/main/java/service/AuthService.java
package service;

import domain.User;
import lombok.Getter;
import repository.UsersDBRepository;
import repository.UsersRepository;
import utils.JdbcUtils;

import javax.crypto.Cipher;
import javax.crypto.spec.SecretKeySpec;
import java.io.IOException;
import java.util.Base64;
import java.util.Optional;
import java.util.Properties;

public class AuthService {
    @Getter
    private static final AuthService instance = new AuthService();
    private final UsersRepository repository;
    private final String encryptionAESKey;

    private AuthService() {
        Properties props = new Properties();

        try {
            props.load(AuthService.class.getClassLoader().getResourceAsStream("db.properties"));
        } catch (IOException e) {
            System.out.println("Cannot find db.properties " + e);
        }

        JdbcUtils jdbcUtils = new JdbcUtils(props);
        repository = new UsersDBRepository(jdbcUtils);

        encryptionAESKey = props.getProperty("jdbc.eckey");
    }

    public User login(String email, String password) {
        String encryptedPassword = encrypt(password, encryptionAESKey);

        Optional<User> user = repository.findByEmailAndPassword(email, encryptedPassword);

        if(user.isEmpty())
            throw new RuntimeException("Invalid credentials");

        return user.get();
    }

    public void register(String email, String password) {
        String encryptedPassword = encrypt(password, encryptionAESKey);

        Optional<User> user = repository.findByEmailAndPassword(email, encryptedPassword);

        if(user.isPresent())
            throw new RuntimeException("Email already exists");

        repository.save(new User(0L, email, encryptedPassword));
    }

    public String encrypt(String data, String key) {
        SecretKeySpec secretKey = new SecretKeySpec(key.getBytes(), "AES");

        Cipher cipher = null;
        try {
            cipher = Cipher.getInstance("AES/ECB/PKCS5Padding");
            cipher.init(Cipher.ENCRYPT_MODE, secretKey);

            byte[] encryptedBytes = cipher.doFinal(data.getBytes());
            return Base64.getEncoder().encodeToString(encryptedBytes);
        }
        catch (Exception exception) {
            System.out.println(exception);
            throw new RuntimeException("Password encryption failed!");
        }
    }
}
